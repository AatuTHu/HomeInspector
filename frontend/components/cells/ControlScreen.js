import { View, Text, ScrollView} from "react-native"
import { styles } from "../../styles/styles"
import  localVariables  from "../../env"
import { useState } from "react"
import Buttons from "../atoms/Buttons"
import SaveSensorLocation from "../molecules/SaveSensorLocation"
import DropDown  from "../molecules/DropDown"
import StartMeasuring from "../molecules/StartMeasuring"
import LightSwitch from "../molecules/LightSwitch"

export default function ControlScreen({setScreen,setTempStarted,tempStarted,setCurrentHumLoc,setCurrentTempLoc,currentHumLoc,currentTempLoc,setHumStarted,humStarted,setLights, lights}) {

  const options = [
    {id : 1 ,name : "humidity", value : "humidity"},
    {id : 2, name : "temperature", value : "temperature"}
  ]

  const [selected, setSelected] = useState("")
  
  const showAll = () => { 
    setScreen(4);
  }

  return (
    <ScrollView>
      <View style = { styles.controlContainer }>
          <DropDown selected={selected} setSelected={setSelected} options={options}/>
          { selected ? (<>
            <SaveSensorLocation 
              selected = {selected} 
              setCurrentHumLoc = {setCurrentHumLoc} 
              setCurrentTempLoc = {setCurrentTempLoc} 
              currentHumLoc = {currentHumLoc} 
              currentTempLoc = {currentTempLoc}
            />
            <StartMeasuring 
              selected = {selected} 
              setTempStarted = { setTempStarted } 
              tempStarted = {tempStarted} 
              setHumStarted = { setHumStarted } 
              humStarted = {humStarted}
            />
          </>) : (<></>)}
       
          <LightSwitch lights = {lights} setLights = {setLights}/>
          <Buttons text = "See all the measurements" type = "update" onPress={ showAll }/>
      </View>
    </ScrollView>
  )
}