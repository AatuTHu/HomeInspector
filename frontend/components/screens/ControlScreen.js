import { View, Text, ScrollView} from "react-native"
import  variables  from "../../env"
import { styles } from "../../styles/styles"
import { useState } from "react"
import Buttons from "./atoms/Buttons"
import InfoBox from "./atoms/InfoBox"
import InputBox from "./atoms/InputBox"


export default function Control({ssid, location, setScreen}) {

  const { apiKey,locationURL,lightURL,startURL } = variables

  const [newLocation, setNewLocation] = useState("");
  const [statusText, setStatusText] = useState("")
  const [timer, setTimer] = useState(0)
  const [start, setStart] = useState(false)
  const [lights, setLights] = useState(false)

  const lightSwitch = async() => {
    setLights(!lights)
    let mode = lights ?  0 : 1
    await fetch(lightURL, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: `apiKey=${apiKey}&lightMode=${mode}`,
    }).then( res => {
      if(res.status === 200) {
        lights ? setStatusText("Lights Off") : setStatusText("Lights On")
      } else {
        setStatusText("There was an error, Try again later")
        setLights(false)
      }
    }).catch( error => {
      console.log(error)
    })
  }

  const saveLocation = async() => {
   await fetch(locationURL, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: `apiKey=${apiKey}&location=${newLocation}`,
    }).then( res => {
      if(res.status === 200) {
        setStatusText("Location switched")
      } else {
        setStatusText("There was an error, Try again later")
      }
    }).catch( error => {
      console.log(error)
    })
  }

  const startMeasuring = async() => {
    setStart(!start)
    let mode = start? 0 : 1
    await fetch(startURL, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: `apiKey=${apiKey}&measuringMode=${mode}`,
    }).then( res => {
      if(res.status === 200) {
        start ? setStatusText("Measurements stopped") : setStatusText("Measurements started")
      } else {
        setStatusText("There was an error, Try again later")
      }
    }).catch( error => {
      console.log(error)
    })
  }

  const showAll = () => { 
    setScreen(4);
  }

  const deleteAll = () => {
    
  }

  return (
    <>
    <ScrollView>
      <View style = { styles.controlContainer }>
        <InfoBox text = "Wi-Fi ssid" textStyle = 'headingText' type = 'heading'/>
          <InfoBox text = { ssid } type =  'small' textStyle = 'subHeadingText' />
            <InfoBox text = "Device location" textStyle = 'headingText' type = 'heading'/>  
          <InputBox placeholder= "where ..." text = { location } newText = { setNewLocation } onSubmitEditing={ saveLocation }/>
        <Buttons text = "save" type = "save" onPress={ saveLocation } />

      <View style = { styles.block}>
        <Text style = { styles.statusText }>{statusText}</Text>
      </View>

      <View style = { styles.buttonRowContainer }>
        <Buttons text = "start" type = "start" onPress={ startMeasuring }/>
          <InfoBox text = { timer } type = 'status' textStyle = 'subHeadingText' />
      </View>

      <View style = { styles.buttonRowContainer }>
        <Buttons text = "Lights on/off" type = "start" onPress={ lightSwitch }/>
        <View style = { styles.statusBox }>
            <Text style = { styles.subHeadingText }>{lights ? (<Text style = {{"color" : "yellow"}} >On</Text>) : (<Text>Off</Text>)}</Text>
        </View>
      </View>

          <Buttons text = "See all the measurements" type = "update" onPress={ showAll }/>
        <Buttons text = "Delete measurements" type = "delete" onPress={ deleteAll }/>
      </View>
    </ScrollView>
  </>
  )
}