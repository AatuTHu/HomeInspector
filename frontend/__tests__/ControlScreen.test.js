import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import ControlScreen from '../components/cells/ControlScreen';


describe('ControlScreen', () => {
  
  it('renders correctly', () => {
    const screen = render(<ControlScreen/>);
    expect(screen).toMatchSnapshot();
  });

  it('should change screen if all measurements button is pressed', () => {
    const setScreen = jest.fn();
    const {getByText} = render(<ControlScreen setScreen={setScreen}/>);

    const changeScreen = getByText('See all the measurements');
    fireEvent.press(changeScreen)

    expect(setScreen).toHaveBeenCalledTimes(1);
  })

  it('should turn on the lights', async() => {
    global.fetch = jest.fn(() =>
        Promise.resolve({
          status: 200,
          json: () => Promise.resolve({}),
    }));
      setLights = jest.fn()
      const { getByText } = render(<ControlScreen lights={true} setLights={setLights} />);
      fireEvent.press(getByText('lights'));
      await waitFor(() => 
        expect(getByText('On')).toBeTruthy(),
      )
  })

  it('should turn off the lights', async() => {
    global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 304,
      json: () => Promise.resolve({}),
    }));
    setLights = jest.fn()
    const { getByText } = render(<ControlScreen lights={false} setLights={setLights} />);
    fireEvent.press(getByText('lights'));
    await waitFor(() => 
      expect(getByText('Off')).toBeTruthy(),
    )
    })

    it('should show option when select a sensor is pressed', () => {
      const { getByText } = render(<ControlScreen/>)
      const selectButton = getByText('Select a sensor..')

      fireEvent.press(selectButton)

      expect(getByText('humidity')).toBeTruthy();
      expect(getByText('temperature')).toBeTruthy();
    })

    it('should be able to choose a sensor from dropdown', () => {
      const { getByText, getByPlaceholderText } = render(<ControlScreen/>)
      const selectButton = getByText('Select a sensor..')

      fireEvent.press(selectButton)
      
      const humiditySensor = getByText('humidity')

      fireEvent.press(humiditySensor)

      expect(getByPlaceholderText('update location')).toBeTruthy()
      expect(getByText('Save')).toBeTruthy()
      expect(getByText('Start')).toBeTruthy()
    })

    it('should be able to save new location when selected humidity', async () => {
      global.fetch = jest.fn(() =>
        Promise.resolve({
          status: 202,
          json: () => Promise.resolve({}),
        }));
      
    const { getByText } = render(<ControlScreen setCurrentHumLoc={jest.fn()}/>);
    const selectButton = getByText('Select a sensor..')
     fireEvent.press(selectButton)
    const humiditySensor = getByText('humidity')
     fireEvent.press(humiditySensor)    
    const saveButton = getByText('Save');
     fireEvent.press(saveButton);

    await waitFor(() => expect(getByText('Location saved')).toBeTruthy())
  })

  it('should be able to save new location when selected temperature', async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        status: 202,
        json: () => Promise.resolve({}),
    }));
    
    const { getByText } = render(<ControlScreen setCurrentTempLoc={jest.fn()}/>);
    const selectButton = getByText('Select a sensor..')
     fireEvent.press(selectButton)
    const TemperatureSensor = getByText('temperature')
     fireEvent.press(TemperatureSensor)    
    const saveButton = getByText('Save');
     fireEvent.press(saveButton);

    await waitFor(() => expect(getByText('Location saved')).toBeTruthy())
  })

  it('should be able to start measuring', async() => {
    global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({}),
    }));
    
    const setHumStarted = jest.fn()
    const { getByText } = render(<ControlScreen setHumStarted={setHumStarted} humStarted={false}/>);
    const selectButton = getByText('Select a sensor..')
     fireEvent.press(selectButton)
    const humiditySensor = getByText('humidity')
     fireEvent.press(humiditySensor)
    const startButton = getByText('Start');
     fireEvent.press(startButton)

     await waitFor(() => expect(setHumStarted).toHaveBeenCalledWith(true));
  })

});