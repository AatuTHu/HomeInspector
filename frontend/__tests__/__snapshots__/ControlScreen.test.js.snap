// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlScreen renders correctly 1`] = `
<RCTScrollView>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          {
            "borderRadius": 22,
            "height": 50,
            "justifyContent": "center",
            "marginTop": 20,
            "width": "95%",
          }
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 20,
                "fontWeight": "bold",
              },
            ]
          }
        >
          Sensor
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            {
              "color": "#001021",
              "fontSize": 22,
              "textAlign": "center",
            }
          }
        >
          Select a sensor..
        </Text>
      </View>
      <View
        style={
          {
            "backgroundColor": "#FFFF",
            "flex": 1,
            "flexDirection": "column",
            "width": "100%",
          }
        }
      >
        <View>
          <Text
            style={
              {
                "color": "black",
                "fontSize": 20,
                "opacity": 0.5,
                "padding": 10,
                "textAlign": "center",
              }
            }
          />
        </View>
      </View>
      <View
        style={
          {
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 20,
            "marginTop": 1,
            "width": "95%",
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#03254E",
              "borderRadius": 20,
              "borderWidth": 1,
              "opacity": 1,
              "padding": 5,
              "width": "50%",
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 22,
                "textAlign": "center",
              }
            }
          >
            lights
          </Text>
        </View>
        <View
          style={
            {
              "backgroundColor": "#D9E5D6",
              "borderRadius": 22,
              "justifyContent": "center",
              "width": "30%",
            }
          }
        >
          <Text
            style={
              {
                "fontSize": 22,
                "textAlign": "center",
              }
            }
          >
            <Text>
              Off
            </Text>
          </Text>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#03254E",
            "borderRadius": 20,
            "borderWidth": 1,
            "color": "white",
            "marginBottom": 10,
            "marginTop": 20,
            "opacity": 1,
            "padding": 10,
            "width": "95%",
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontSize": 22,
              "textAlign": "center",
            }
          }
        >
          See all the measurements
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
